### PR TITLE
Prompts from file support nodes

### PR DESCRIPTION
**Pull request by skunkworxdark**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission

## Have you discussed this change with the InvokeAI team?
- [ ] Yes
- [X] No, because:  I have messaged in the dev-chat but no answer ATM.  I thought it easier to create the PR and the edit or delete it if not needed or the wrong way to do it. 
      
## Have you updated all relevant documentation?
- [ ] Yes
- [X] No

## Description
Proposed 4 support nodes for the existing promptsFromFile node:

- PromptSplitNeg - This would take input in the old 2.3.5 format "Positive [Negative]" and split it into two separate prompts everything inside [] would go to the neg prompt and everything else would go to the positive.  This would allow the PromptsFromFile to contain both positive and negative prompts on a single line. This is the class name I like the least as it really doesn't split in the normal programming sense its more of an extract but split is an easy concept to get for users.
- PromptJoin - This would take two prompts and join them into one. The use case I have in mind is that you could add to the beginning or end of individual split prompts. I toyed with calling it PromptConcatenate but feel it is a bit to programming orientated for the average user.
- PromptReplace - This would take 3 inputs (a prompt, a search string and a replace string).  The use case I have in mind for this is the PromptsFromFile would contain multiple template like prompts (e.g "photo of XXXXXX , Nikon", "drawing of XXXXXX in crayon", "Wikihow of XXXXXX, in a cartoon style") with a fixed place holder ( "XXXXXX") that would be replaced for each line of the file with whatever you put tin the replace field. This way you could change the subject of each prompt and try it against multiple templates.  Or just remove terms from prompts, I have used it to remove embeddings from prompts for SDXL (e.g.<neg-sketch-2>).  This node by default is case insensitive, it also has a boolean toggle to allow switching the a regex style search for power users.
- PromptsToFile -  This allows saving of prompt strings to a file from within the node graph.  With this node you can keep a track of all the prompts used in a session. Will accept a single prompt or a collection of prompts. Has a toggle for file append or overwrite.  Especially useful if the prompts are automatically generated by dynamic prompts, nightmare prompt gen or generative Grammar-based prompt gen. etc.. 

## Related Tickets & Documents

- Related Issue #
- Closes #

## QA Instructions, Screenshots, Recordings

Some node graph examples of use
image
image
image
image

## Added/updated tests?

- [ ] Yes
- [X] No : _please replace this line with details on why tests
      have not been included_

### CODE REVIEW

File: invokeai/app/invocations/prompt_tools.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+
+    prompt_left: str = InputField(default="", description="Prompt Left", ui_component=UIComponent.Textarea)
+    prompt_right: str = InputField(default="", description="Prompt Right", ui_component=UIComponent.Textarea)
+
+    def invoke(self, context: InvocationContext) -> StringOutput:
+        return StringOutput(value=((self.prompt_left or "") + (self.prompt_right or "")))
+
+
+@invocation("prompt_join_three", title="Prompt Join Three", tags=["prompt", "join"], category="prompt", version="1.0.0")
+class PromptJoinThreeInvocation(BaseInvocation):
```
Comment (psychedelicious) [inline invokeai/app/invocations/prompt_tools.py:109]: I'd like to make this more generalized... it would be nice if there was just a single extensible string concat node that let you concat any number of strings, but that will need to come in a future change. For now, this is fine

File: invokeai/app/invocations/prompt_tools.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+
+    def invoke(self, context: InvocationContext) -> StringOutput:
+        return StringOutput(value=((self.prompt_left or "") + (self.prompt_middle or "") + (self.prompt_right or "")))
+
+
+@invocation(
+    "prompt_replace", title="Prompt Replace", tags=["prompt", "replace", "regex"], category="prompt", version="1.0.0"
+)
+class PromptReplaceInvocation(BaseInvocation):
+    """Replaces the search string with the replace string in the prompt"""
```
Comment (psychedelicious) [inline invokeai/app/invocations/prompt_tools.py:124]: Should just be a generic StringRegexReplaceInvocation.

File: invokeai/app/invocations/prompt_tools.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+    height: int
+    steps: int
+    cfg_scale: float
+
+
+@invocation_output("pt_fields_collect_output")
+class PTFieldsCollectOutput(BaseInvocationOutput):
+    """PTFieldsCollect Output"""
+
+    pt_fields: str = OutputField(description="PTFields in Json Format")
```
Comment (psychedelicious) [inline invokeai/app/invocations/prompt_tools.py:164]: Can you walk me through why this needs to be a string representation? Why can't we leave this as a Pydantic model?

I see some gymnastics in the various nodes to facilitate parsing the data from string to pydantic model - but pydantic already handles serialization and deserialization...

Comment (psychedelicious) [inline invokeai/app/invocations/prompt_tools.py:164]: Can you please explain the use case for the `PTFields` model? It seems to be used as intermediate storage of a select number of fields. Why not use a primitive for each? I'm wary of using a new data structure instead of composing existing ones

File: invokeai/app/invocations/prompt_tools.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+                    positive_prompt=self.positive_prompt,
+                    positive_style_prompt=self.positive_style_prompt,
+                    negative_prompt=self.negative_prompt,
+                    negative_style_prompt=self.negative_style_prompt,
+                    seed=self.seed,
+                    width=self.width,
+                    height=self.height,
+                    steps=self.steps,
+                    cfg_scale=self.cfg_scale,
+                ).dict()
```
Comment (psychedelicious) [inline invokeai/app/invocations/prompt_tools.py:202]: All of these parameters are marked as Optional, but `PTFields` requires them. Maybe `PTFields` needs default values

File: invokeai/app/invocations/prompt_tools.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+
+@invocation(
+    "prompt_split_neg",
+    title="Prompt Split Negative",
+    tags=["prompt", "split", "negative"],
+    category="prompt",
+    version="1.0.0",
+)
+class PromptSplitNegInvocation(BaseInvocation):
+    """Splits prompt into two prompts, inside [] goes into negative prompt everthing else goes into positive prompt. Each [ and ] character is replaced with a space"""
```
Comment (psychedelicious) [inline invokeai/app/invocations/prompt_tools.py:66]: I wonder if this could be better as a generic string operation that splits a string on a specific string. 

In that case, the prompt could be like "my positive prompt stuff || my negative prompt stuff" and you'd tell the node to split on "||".

I understand this goes against the convention of nesting the negative prompt within `[...]` but parsing nested stuff is never fun and often has edge cases. I can't understand the logic here without stepping through it line by line - the intent isn't quite clear.

A simple split is much easier to reason about than the logic here, which I'm sure works, but which I cannot really wrap my head around without stepping through it line by line.

Anyways, just thinking out loud

**Review by psychedelicious — CHANGES_REQUESTED** [review]

File: invokeai/app/invocations/prompt_tools.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+
+@invocation("prompt_join", title="Prompt Join", tags=["prompt", "join"], category="prompt", version="1.0.0")
+class PromptJoinInvocation(BaseInvocation):
+    """Joins prompt left to prompt right"""
+
+    prompt_left: str = InputField(default="", description="Prompt Left", ui_component=UIComponent.Textarea)
+    prompt_right: str = InputField(default="", description="Prompt Right", ui_component=UIComponent.Textarea)
+
+    def invoke(self, context: InvocationContext) -> StringOutput:
+        return StringOutput(value=((self.prompt_left or "") + (self.prompt_right or "")))
```
[the remark this answers is not in the document]
Comment (skunkworxdark) [reply invokeai/app/invocations/prompt_tools.py:105]: I have no issue with the change of name or moving it into a strings.py file. I debated the naming when I originally wrote it  but thought concatenate string was a bit to much programmer speak and stuck to the more use based way of naming as I thought this fitted in with the existing naming in invoke at the time.

File: invokeai/app/invocations/prompt_tools.py (hunk 1 above)
Comment (skunkworxdark) [reply invokeai/app/invocations/prompt_tools.py:109]: I totally agree an extensible version would be the ideal and would be very useful in many other nodes.  Additionally this functionality can be recreated using the existing single join version but I found that joining 3 things happened enough times that a separate node just made sense.    I assume this should also go into strings.py

File: invokeai/app/invocations/prompt_tools.py (hunk 2 above)
Comment (skunkworxdark) [reply invokeai/app/invocations/prompt_tools.py:124]: agree if the others are moving the name from prompt to string then this should be the same and also move into strings.py

File: invokeai/app/invocations/prompt_tools.py (hunk 3 above)
Comment (skunkworxdark) [reply invokeai/app/invocations/prompt_tools.py:164]: The idea of the PTField nodes is to use them in conjunction with the PromptsToFile and PromptsFromFile nodes to allow loading and saving of more than just the prompts but also all the other parameters for generating an image.  This is much more a proof of concept idea than a finished node as not all parameters at the moment can be collected (e.g. model, vae, loras used etc) as the current node architecture doe not expose them as an output .

I originally tried to using a custom type as the output type and was going to write versions of the ToFile and FromFile that would accept them but I just couldn't get it to work. From memory (which is a bit poor nowadays as I suffer from CFS) the frontend just didn't like the custom field without changes to UI side of things which I felt where out of scope of a community node.    However I think this worked out well as converting to a json string allowed using the existing PromptsToFile and PromptsFromFile Nodes without any changes.  Also it means the text files ended up relatively human readable and editable.   I hope that makes sense and I have understood your question.

Comment (skunkworxdark) [reply invokeai/app/invocations/prompt_tools.py:164]: See Previous comment as I think I answer this there.

File: invokeai/app/invocations/prompt_tools.py (hunk 4 above)
Comment (skunkworxdark) [reply invokeai/app/invocations/prompt_tools.py:202]: The idea of them been optional was not cast in stone as this was still just proof of concept for the moment.   My thought process was that you could pick and choose which parameters to pull in from a file or save to a file and then override them if needed but if the parameter was not in the file then it would get a default value that was entered by the user in the node.  I don't think that currently this is achievable in a easy way with the UI elements we have available in nodes at the moment.

File: invokeai/app/invocations/prompt_tools.py (hunk 5 above)
Comment (skunkworxdark) [reply invokeai/app/invocations/prompt_tools.py:66]: I think a generic string split would be a great new node and would have many use cases and should be part of a new strings.py 

However this node came in as a bridging idea between the old Invoke and new invoke as I had many old style prompts and wanted to test them against different models and parameters.  I agree that nesting can and does have many edge cases and I apologise if my code is not as clear in intent as is could be but it did grow in an organic manner as users of it asked for additional bits for it. Originally it was almost just a simple split but people wanted to do thing like combine multiple prompts and still have the neg side of it pulled out. So it just grew in complexity.

Comment (skunkworxdark) [reply invokeai/app/invocations/prompt_tools.py:66]: Just so I understand what is going on here.  Did you want me to make these name changes and move them into a new strings.py file?    What would you like to do with the PromptsToFileInvocation?  I am thinking that this should also become StringsToFile and the existing core version (prompts.py) of PromptsFromFileInvocation should also move and be renamed.   What's your thoughts?

File: invokeai/app/invocations/prompt_tools.py (hunk 3 above)
Comment (skunkworxdark) [reply invokeai/app/invocations/prompt_tools.py:164]: To be clear the PTField nodes where only ever a proof of concept and I don't think they should be in core in its current form as it is far from complete. Without the core nodes framework exposing the additional fields used in generating an image then it will never be fully useful.   For it to be fully useful then it needs to be extensible in the way you can just add fields to the node that you want to save and do the opposite when loading them back from a file.

**Comment by skunkworxdark** [comment]
Just so I understand what is expected, as I have very little experience of working in an open source project especially one  as large as this and I would just like to be more of a help than a hindrance to the project. So I have a couple of questions:

- Did you want me to make these name changes and move them into a new strings.py file?

- What would you like to do with the new PromptsToFileInvocation and the existing core version (prompts.py) of PromptsFromFileInvocation. Do we also rename these to Strings in strings.py or keep them as Prompts in the Prompts.py file?

- Finally any thoughts on the PromptStrengthInvocation and the PromptStrengthsCombineInvocation nodes as I don't think they really fit nicely into string manipulation. Prombably more into the prompts side of things or maybe not even into core at all?

**Comment by skunkworxdark** [comment]
In an attempt to speed things up as we are in different time zones I have gone ahead with the recommended changes anyway. 

I moved the nodes to strings.py and renamed them to String related rather than Prompt related.
I also moved and renamed the existing PromptFromFile (from prompt.py) into strings.py

The PTFields remain in the prompt_tools.py until a decision is made on them.  I favour not adding them to core at the moment as I don't think they are complete enough or fully useful in the current state.

File: invokeai/app/invocations/prompt_tools.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (skunkworxdark) [reply invokeai/app/invocations/prompt_tools.py]: Name changed to StringJoin

File: invokeai/app/invocations/prompt_tools.py (hunk 2 above)
Comment (skunkworxdark) [reply invokeai/app/invocations/prompt_tools.py]: Name changed to StringReplace and moved into strings.py . Didn't put regex in the name as that is an option that is off as default and might confuse some users.

File: invokeai/app/invocations/prompt_tools.py (hunk 3 above)
Comment (skunkworxdark) [reply invokeai/app/invocations/prompt_tools.py]: These have been removed from from the PR and will live in my own community custom nodes for now.

File: invokeai/app/invocations/prompt_tools.py (hunk 4 above)
Comment (skunkworxdark) [reply invokeai/app/invocations/prompt_tools.py]: same as the other PTfield nodes

File: invokeai/app/invocations/prompt_tools.py (hunk 5 above)
Comment (skunkworxdark) [reply invokeai/app/invocations/prompt_tools.py]: New node StringSplit added that will split a string into two strings based upon the first occurance of the delimiter
Changed the name of the PromptSplitNegative to StringsplitNegative and moved into strings.py

**Comment by skunkworxdark** [comment]
In summary this PR now has these changes.

StringsFromFile - renamed and moved from prompt.py to strings.py
StringsToFile, StringJoin, StringJoinThree, StringReplace, StringSplitNeg, StringWeight, StringWeightsCombine - renamed and moved to strings.py
StringSplit - New generic StringSplit that Splits a string into two based upon a supplied delimiter on its first occurrence

**Comment by skunkworxdark** [comment]
Finally there was some discussion on discord over the validity of the StringWeight, StringWeightsCombine nodes inclusion into core as it was thought that string manipulation of weight been fed into a compel node was the wrong approach to achieve .and() and .blend() and this should be done at a conditioning level instead.   So some guidance from the core team would be useful over these two nodes and if anything else needs to be changed about the other ones.  Then I can either take them out or keep them in the PR.

**Comment by hipsterusername** [comment]
Will try to collect some thoughts here, but pinging to see if @psychedelicious or @damian0815 have any

**Comment by Millu** [comment]
I think what makes sense to me here is adding string split & prompts from file to core, and leaving StringWeightsCombine & StringWeight as Community Nodes with out-the-box Invoke weighting being down through .blend(). This will help maintain continuity in how prompts are treated throughout Invoke, reducing possible paths of confusion for users and makes it easier to maintain going forward 

@skunkworxdark thank you so much for bearing with the discussion & back and forth on these nodes! They're obviously of interest to folks which is why they've garnered so many thoughts

**Comment by skunkworxdark** [comment]
@hipsterusername @Millu @psychedelicious
Is this a definitive way forward or just an individual opinion? 

So we can have an agreement here is a summary of what I think is in and out

In core: 
StringsFromFile - renamed and moved from prompt.py to strings.py
StringsToFile
StringJoin
StringJoinThree
StringReplace
StringSplitNeg
StringSplit - New generic StringSplit that Splits a string into two based upon a supplied delimiter on its first occurrence

Not In core but will remain part of my community nodes:
StringWeight
StringWeightsCombine
PTFields nodes

**Comment by hipsterusername** [comment]
> @hipsterusername @Millu @psychedelicious Is this a definitive way forward or just an individual opinion?
> 
> So we can have an agreement here is a summary of what I think is in and out
> 
> In core: StringsFromFile - renamed and moved from prompt.py to strings.py StringsToFile StringJoin StringJoinThree StringReplace StringSplitNeg StringSplit - New generic StringSplit that Splits a string into two based upon a supplied delimiter on its first occurrence
> 
> Not In core but will remain part of my community nodes: StringWeight StringWeightsCombine PTFields nodes
> 
> As soon as I get a definitive yes or no I will make any changes that are needed.

For the sake of giving you some clarity in the near term, let's focus on the areas where there are clear answers for inclusion in the core, and leave the remainder for further discussion/review.

Thanks for understanding the delay in response - Busy times!

In core:
StringJoin
StringJoinThree
StringReplace
StringSplitNeg 
StringSplit - New generic StringSplit that Splits a string into two based upon a supplied delimiter on its first occurrence

Not In core but will remain part of community nodes for the time being:
StringsFromFile 
StringsToFile
StringWeight
StringWeightsCombine
PTFields nodes

The main rationale on moving the file-based stuff down is having "core" support for these would imply more file-path management (Given general patterns elsewhere) than we've currently built node-ecosystem support for. Before we incorporate the node into core, the team needs to have better clarity into other downstream impacts. Hope rationale makes sense!

We'll get these merged pretty quickly once updates are in. Let us know if any feedback!

**Review by psychedelicious — APPROVED** [review]
Sorry for the delay in getting this merged, and thanks for working to get the scope narrowed to fit with the vision for core nodes. I've tested the nodes and fixed a small typo ("spilt -> split").
